routing/endpoint_registry.go

```diff
@@ -0,0 +1,103 @@
+package routing
+
+import (
+	"sync"
+	"time"
+)
+
+// metrics describe the data about endpoint that could be
+// used to perform better load balancing, fadeIn, etc.
+type metrics interface {
+	DetectedTime() time.Time
+	InflightRequests() int64
+}
+
+type entry struct {
+	detected         time.Time
+	inflightRequests int64
+}
+
+var _ metrics = &entry{}
+
+func (e *entry) DetectedTime() time.Time {
+	return e.detected
+}
+
+func (e *entry) InflightRequests() int64 {
+	return e.inflightRequests
+}
+
+type EndpointRegistry struct {
+	mu sync.Mutex
+
+	data map[string]*entry
+}
+
+var _ PostProcessor = &EndpointRegistry{}
+
+type RegistryOptions struct {
+}
+
+func (r *EndpointRegistry) Do(routes []*Route) []*Route {
+	for _, route := range routes {
+		for _, epi := range route.LBEndpoints {
+			metrics := r.GetMetrics(epi.Host)
+			if metrics.DetectedTime().IsZero() {
+				r.SetDetectedTime(epi.Host, time.Now())
+			}
+		}
+	}
+	return routes
+}
+
+func NewEndpointRegistry(o RegistryOptions) *EndpointRegistry {
+	return &EndpointRegistry{data: map[string]*entry{}}
+}
+
+func (r *EndpointRegistry) GetMetrics(key string) metrics {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	e := r.getOrInitEntryLocked(key)
+	copy := &entry{}
+	*copy = *e
+	return copy
```

Comment:
I think you can just `return *e`

Reply:
It is possible, yeah, but does not make sense from the performance point of view. You can see the version with non-pointer receiever on the left and pointer reciever on the right. This benchstat shows that it is better to leave this getter as it is.
```
                                       │ ./epregistry.txt │           ./epregistry_pr.txt           │
                                       │      sec/op      │    sec/op     vs base                   │
GetInflightRequests/1_goroutines-16         147.95n ±  3%   17.52n ±  7%  -88.16% (n=18+10)
GetInflightRequests/2_goroutines-16         181.65n ±  3%   61.22n ± 23%  -66.30% (p=0.000 n=18+10)
GetInflightRequests/3_goroutines-16         213.20n ±  5%   78.99n ±  6%  -62.95% (p=0.000 n=18+10)
GetInflightRequests/4_goroutines-16          224.0n ±  2%   112.9n ± 27%  -49.59% (p=0.000 n=18+10)
GetInflightRequests/5_goroutines-16          237.2n ±  2%   129.5n ± 13%  -45.42% (p=0.000 n=18+10)
GetInflightRequests/6_goroutines-16          241.2n ±  5%   153.0n ± 13%  -36.55% (p=0.000 n=18+10)
GetInflightRequests/7_goroutines-16          255.6n ±  2%   157.6n ± 19%  -38.35% (p=0.000 n=18+10)
GetInflightRequests/8_goroutines-16          264.7n ± 12%   161.9n ± 14%  -38.82% (p=0.000 n=18+10)
GetInflightRequests/12_goroutines-16         297.6n ±  4%   182.0n ± 19%  -38.84% (p=0.000 n=17+10)
GetInflightRequests/16_goroutines-16         312.6n ±  8%   186.0n ±  3%  -40.52% (p=0.000 n=17+10)
GetInflightRequests/24_goroutines-16         335.4n ±  8%   165.3n ± 12%  -50.72% (p=0.000 n=17+10)
GetInflightRequests/32_goroutines-16         337.0n ±  9%   137.7n ± 23%  -59.15% (p=0.000 n=17+10)
GetInflightRequests/48_goroutines-16         308.7n ±  6%   114.8n ± 11%  -62.81% (p=0.000 n=17+10)
GetInflightRequests/64_goroutines-16         299.5n ± 14%   110.4n ±  8%  -63.14% (p=0.000 n=17+10)
GetInflightRequests/128_goroutines-16        299.7n ±  8%   110.4n ±  4%  -63.15% (p=0.000 n=17+10)
GetInflightRequests/256_goroutines-16        304.6n ±  8%   114.0n ±  7%  -62.56% (p=0.000 n=17+10)
GetInflightRequests/512_goroutines-16        303.5n ±  4%   113.0n ±  6%  -62.75% (p=0.000 n=17+10)
GetInflightRequests/768_goroutines-16        314.4n ±  5%   121.4n ±  7%  -61.40% (p=0.000 n=17+10)
GetInflightRequests/1024_goroutines-16       286.0n ± 11%   120.4n ±  9%  -57.92% (p=0.000 n=17+10)
GetInflightRequests/1536_goroutines-16       273.1n ± 12%   116.8n ±  8%  -57.23% (p=0.000 n=17+10)
GetInflightRequests/2048_goroutines-16       247.8n ± 11%   116.3n ±  6%  -53.07% (p=0.000 n=17+9)
GetInflightRequests/4096_goroutines-16       218.7n ±  5%   104.9n ±  8%  -52.03% (p=0.000 n=17+9)
geomean                                      263.3n         112.7n        -57.18%
```

Reply:
I see, you return pointer to a copy.
It makes sense to check allocations in the bechmark as well (via b.ReportAllocs())

Reply:
https://pkg.go.dev/testing#B.ReportAllocs
> It is equivalent to setting -test.benchmem

I have used this setting as well, and that is what I see in benchstat. Pointer receiver seems the best choice from any point of view.

```
                                       │ ./epregistry.txt │          ./epregistry_pr.txt           │
                                       │       B/op       │   B/op     vs base                     │
GetInflightRequests/1_goroutines-16            32.00 ± 0%   0.00 ± 0%  -100.00% (n=18+17)
GetInflightRequests/2_goroutines-16            32.00 ± 0%   0.00 ± 0%  -100.00% (n=18+17)
GetInflightRequests/3_goroutines-16            32.00 ± 0%   0.00 ± 0%  -100.00% (n=18+17)
GetInflightRequests/4_goroutines-16            32.00 ± 0%   0.00 ± 0%  -100.00% (n=18+17)
GetInflightRequests/5_goroutines-16            32.00 ± 0%   0.00 ± 0%  -100.00% (n=18+17)
GetInflightRequests/6_goroutines-16            32.00 ± 0%   0.00 ± 0%  -100.00% (n=18+17)
GetInflightRequests/7_goroutines-16            32.00 ± 0%   0.00 ± 0%  -100.00% (n=18+17)
GetInflightRequests/8_goroutines-16            32.00 ± 0%   0.00 ± 0%  -100.00% (n=18+17)
GetInflightRequests/12_goroutines-16           32.00 ± 0%   0.00 ± 0%  -100.00% (p=0.000 n=17)
GetInflightRequests/16_goroutines-16           32.00 ± 0%   0.00 ± 0%  -100.00% (p=0.000 n=17)
GetInflightRequests/24_goroutines-16           32.00 ± 0%   0.00 ± 0%  -100.00% (p=0.000 n=17)
GetInflightRequests/32_goroutines-16           32.00 ± 0%   0.00 ± 0%  -100.00% (n=17+16)
GetInflightRequests/48_goroutines-16           32.00 ± 0%   0.00 ± 0%  -100.00% (n=17+16)
GetInflightRequests/64_goroutines-16           32.00 ± 0%   0.00 ± 0%  -100.00% (n=17+16)
GetInflightRequests/128_goroutines-16          32.00 ± 0%   0.00 ± 0%  -100.00% (n=17+16)
GetInflightRequests/256_goroutines-16          32.00 ± 0%   0.00 ± 0%  -100.00% (n=17+16)
GetInflightRequests/512_goroutines-16          32.00 ± 0%   0.00 ± 0%  -100.00% (n=17+16)
GetInflightRequests/768_goroutines-16          32.00 ± 0%   0.00 ± 0%  -100.00% (n=17+16)
GetInflightRequests/1024_goroutines-16         32.00 ± 0%   0.00 ± 0%  -100.00% (n=17+16)
GetInflightRequests/1536_goroutines-16         32.00 ± 0%   0.00 ± 0%  -100.00% (n=17+16)
GetInflightRequests/2048_goroutines-16         32.00 ± 0%   0.00 ± 0%  -100.00% (n=17+16)
GetInflightRequests/4096_goroutines-16         32.00 ± 0%   0.00 ± 0%  -100.00% (n=17+16)
geomean                                        32.00                   ?                       ¹ ²
¹ summaries must be >0 to compute geomean
² ratios must be >0 to compute geomean

                                       │ ./epregistry.txt │           ./epregistry_pr.txt           │
                                       │    allocs/op     │ allocs/op   vs base                     │
GetInflightRequests/1_goroutines-16            1.000 ± 0%   0.000 ± 0%  -100.00% (n=18+17)
GetInflightRequests/2_goroutines-16            1.000 ± 0%   0.000 ± 0%  -100.00% (n=18+17)
GetInflightRequests/3_goroutines-16            1.000 ± 0%   0.000 ± 0%  -100.00% (n=18+17)
GetInflightRequests/4_goroutines-16            1.000 ± 0%   0.000 ± 0%  -100.00% (n=18+17)
GetInflightRequests/5_goroutines-16            1.000 ± 0%   0.000 ± 0%  -100.00% (n=18+17)
GetInflightRequests/6_goroutines-16            1.000 ± 0%   0.000 ± 0%  -100.00% (n=18+17)
GetInflightRequests/7_goroutines-16            1.000 ± 0%   0.000 ± 0%  -100.00% (n=18+17)
GetInflightRequests/8_goroutines-16            1.000 ± 0%   0.000 ± 0%  -100.00% (n=18+17)
GetInflightRequests/12_goroutines-16           1.000 ± 0%   0.000 ± 0%  -100.00% (p=0.000 n=17)
GetInflightRequests/16_goroutines-16           1.000 ± 0%   0.000 ± 0%  -100.00% (p=0.000 n=17)
GetInflightRequests/24_goroutines-16           1.000 ± 0%   0.000 ± 0%  -100.00% (p=0.000 n=17)
GetInflightRequests/32_goroutines-16           1.000 ± 0%   0.000 ± 0%  -100.00% (n=17+16)
GetInflightRequests/48_goroutines-16           1.000 ± 0%   0.000 ± 0%  -100.00% (n=17+16)
GetInflightRequests/64_goroutines-16           1.000 ± 0%   0.000 ± 0%  -100.00% (n=17+16)
GetInflightRequests/128_goroutines-16          1.000 ± 0%   0.000 ± 0%  -100.00% (n=17+16)
GetInflightRequests/256_goroutines-16          1.000 ± 0%   0.000 ± 0%  -100.00% (n=17+16)
GetInflightRequests/512_goroutines-16          1.000 ± 0%   0.000 ± 0%  -100.00% (n=17+16)
GetInflightRequests/768_goroutines-16          1.000 ± 0%   0.000 ± 0%  -100.00% (n=17+16)
GetInflightRequests/1024_goroutines-16         1.000 ± 0%   0.000 ± 0%  -100.00% (n=17+16)
GetInflightRequests/1536_goroutines-16         1.000 ± 0%   0.000 ± 0%  -100.00% (n=17+16)
GetInflightRequests/2048_goroutines-16         1.000 ± 0%   0.000 ± 0%  -100.00% (n=17+16)
GetInflightRequests/4096_goroutines-16         1.000 ± 0%   0.000 ± 0%  -100.00% (n=17+16)
geomean                                        1.000                    ?                       ¹ ²
¹ summaries must be >0 to compute geomean
² ratios must be >0 to compute geomean
```